src/dataset-builder/CohortEditor.ts

```diff
@@ -206,75 +229,105 @@ type AddCriteriaSelectorProps = {
-        ),
-      },
-      {
-        label: 'Program Data',
-        options: _.map((programDataOption) => {
-          return {
-            value: programDataOption,
-            label: programDataOption.name,
-          };
-        }, snapshotBuilderSettings.programDataOptions),
-      },
-    ],
-    'aria-label': addCriteriaText,
-    placeholder: addCriteriaText,
-    value: null,
-    onChange: (x) => {
-      if (x !== null) {
-        if (x.value.kind === 'domain') {
-          onStateChange(domainCriteriaSelectorState.new(cohort, criteriaGroup, x.value));
-        } else {
-          const criteria = criteriaFromOption(x.value);
-          if (criteria !== undefined) {
-            updateCohort(_.set(`criteriaGroups.${index}.criteria.${criteriaGroup.criteria.length}`, criteria));
+  const {
+    index,
+    criteriaGroup,
+    updateCohort,
+    dataset: { id: datasetId, snapshotBuilderSettings },
+    onStateChange,
+    cohort,
+  } = props;
+  return (
+    snapshotBuilderSettings &&
+    h(GroupedSelect<CriteriaOption>, {
+      styles: { container: (provided) => ({ ...provided, width: '230px', marginTop: wideMargin }) },
+      isClearable: false,
+      isSearchable: false,
+      options: [
+        {
+          label: 'Domains',
+          options: _.map(
+            (domainOption) => ({
+              value: addKindToDomainOption(domainOption),
+              label: domainOption.category,
+            }),
+            snapshotBuilderSettings.domainOptions
+          ),
+        },
+        {
+          label: 'Program Data',
+          options: _.map((programDataOption) => {
+            return {
+              value: programDataOption,
+              label: programDataOption.name,
+            };
+          }, snapshotBuilderSettings.programDataOptions),
+        },
+      ],
+      'aria-label': addCriteriaText,
+      placeholder: addCriteriaText,
+      value: null,
+      onChange: async (x) => {
+        if (x !== null) {
+          if (x.value.kind === 'domain') {
+            onStateChange(domainCriteriaSelectorState.new(cohort, criteriaGroup, x.value));
+          } else {
+            const criteriaIndex = getNextCriteriaIndex();
+            updateCohort(
+              _.set(`criteriaGroups.${index}.criteria.${criteriaGroup.criteria.length}`, {
+                loading: true,
+                index: criteriaIndex,
+              })
+            );
+            const loadedCriteria = await criteriaFromOption(datasetId, criteriaIndex, x.value);
+
+            if (loadedCriteria !== undefined) {
+              updateCohort((cohort) => {
+                const loading = _.findIndex(
+                  (criteria) => criteria.index === criteriaIndex,
+                  cohort.criteriaGroups[index].criteria
+                );
+                if (loading >= 0) {
+                  return _.set(`criteriaGroups.${index}.criteria.${loading}`, loadedCriteria, cohort);
+                }
+                return cohort;
```

Comment:
So, starting from line 281, the cohort is updated to be loading, then the criteria is loaded, then the cohort is updated again with the loaded criteria.  But I'm a little confused about working of the second update cohort. Specifically, I don't understand the findIndex method.

Reply:
The code creates a new unique index for each criteria (using `getNextCriteriaIndex()`) and then appends a placeholder with the index into the criteria group (above on line 283). Once the criteria has loaded, it uses `findIndex()` to look up the actual index in the criteria group by matching on the generated criteria index. This handles the case where the group index may have changed between where the placeholder was inserted due to other insertions or deletions that have occurred while the loading is taking place.

Maybe calling the generated critiera id an `index` is confusing things here, if so I can rename it to something else.